File: blatt3/scripts/scan_to_cloud.py
#!/usr/bin/env python 

import rospy
import numpy as np
from sensor_msgs.msg import PointCloud, LaserScan
from geometry_msgs.msg import Point32

pub = rospy.Publisher('/cloud', PointCloud, queue_size=10)
filter_size = 5


def laser_callback(scan):
    cloud = PointCloud()
    cloud.header = scan.header

    cloud.header = scan.header;
    for index, point in enumerate(scan.ranges):
        x = point * np.cos(scan.angle_min + index * scan.angle_increment)
        y = point * np.sin(scan.angle_min + index * scan.angle_increment)
        cloud.points.append(Point32(x, y, 0))

    pub.publish(cloud)


def node():
    rospy.init_node('listener', anonymous=True)
    rospy.Subscriber("/scan", LaserScan, laser_callback)
    rospy.spin()


if __name__ == '__main__':
    try:
        node()
    except rospy.ROSInterruptException:
        pass

File: blatt3/src/scan_to_cloud.cpp
#include <ros/ros.h>
#include <sensor_msgs/LaserScan.h>
#include <sensor_msgs/PointCloud.h>
#include <cmath>

ros::Publisher cloud_pub;
int size;

void laserCallback(const sensor_msgs::LaserScan::ConstPtr& scan)
{
  // create a new point cloud
  sensor_msgs::PointCloud cloud;
  cloud.header = scan->header;
  cloud.points.resize(scan->ranges.size());

  // fill the cloud
  for(size_t i = 0; i < scan->ranges.size(); i++)
  {
    cloud.points[i].x = scan->ranges[i] * cos(scan->angle_min + i * scan->angle_increment);
    cloud.points[i].y = scan->ranges[i] * sin(scan->angle_min + i * scan->angle_increment);
  }

  cloud_pub.publish(cloud);
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "scan_to_cloud");
  ros::NodeHandle n;

  ros::Subscriber sub = n.subscribe("/scan", 1, laserCallback);
  cloud_pub = n.advertise<sensor_msgs::PointCloud>("/cloud", 1);

  ros::spin();
  return 0;
}